allennlp/models/multitask.py

```diff
@@ -0,0 +1,196 @@
+from collections import defaultdict
+import inspect
+from typing import Any, Dict, List, Set
+
+from overrides import overrides
+import torch
+
+from allennlp.data import Vocabulary
+from allennlp.modules import Backbone
+from allennlp.models.model import Model
+from allennlp.models.heads import Head
+from allennlp.nn import InitializerApplicator
+
+
+def get_forward_arguments(module: torch.nn.Module) -> Set[str]:
+    signature = inspect.signature(module.forward)
+    return set([arg for arg in signature.parameters if arg != "self"])
+
+
+@Model.register("multitask")
+class MultiTaskModel(Model):
+    """
+    A `MultiTaskModel` consists of a `Backbone` that encodes its inputs in some way, then a
+    collection of `Heads` that make predictions from the backbone-encoded inputs.  The predictions
+    of each `Head` are combined to compute a joint loss, which is then used for training.
+
+    This model works by taking `**kwargs` in `forward`, and passing the right arguments from that to
+    the backbone and to each head.  By default, we use `inspect` to try to figure out getting the
+    right arguments to the right modules, but we allow you to specify these arguments yourself in
+    case our inference code gets it wrong.
+
+    It the caller's responsibility to make sure that the backbone and all heads are compatible with
+    each other, and with the input data that comes from a `MultiTaskDatasetReader`.  We give some
+    arguments in this class and in `MultiTaskDatasetReader` to help with plumbing the arguments in
+    complex cases (e.g., you can change argument names so that they match what the backbone and
+    heads expect).
+
+    # Parameters
+
+    vocab: `Vocab`
+    backbone: `Backbone`
+    heads: `Dict[str, Head]`
+    loss_weights: `Dict[str, float]`, optional (default = `equal weighting`)
+        If you want, you can specify a weight for each head, which we will multiply the loss by when
+        aggregating across heads.  This is equivalent in many cases to specifying a separate
+        learning rate per head, and just putting a weighting on the loss is much easier than
+        figuring out the right way to specify that in the optimizer.
```

Comment:
Doing multi-task in this way doesn't break some important assumptions that optimizers make, does it?

For example, what happens to the momentum of parameters that aren't involved at all for a few batches?

Reply:
If the gradient is `None`, the momentum computation is skipped (I just checked what `Adam` does, and I assume the others do the same).  So this should do the right thing.